### DISCUSSION TITLE
What's the difference between createCookie and createCookieSessionStorage?

### DISCUSSION THREAD

**Discussion by juniorforlife** [remix-run/remix#4931] [Q&A]
As I understand, the different between session & cookie is: 
- Sessions are created by the server. It could be stored locally on a server or a database. 
- The server returns the session ID to the client via cookie. 
- Cookies are stored on client's machine. The browser will attach the cookie to each request header sent to the server.

However, the doc says: 

> For purely cookie-based sessions (where the session data itself is stored in the session cookie with the browser, see [cookies](https://remix.run/docs/en/v1/utils/cookies)) you can use createCookieSessionStorage().

> The main advantage of cookie session storage is that you don't need any additional backend services or databases to use it

Does this mean there's no session being stored on the server? If so then how is this any different from a cookie created by `createCookie`

**Comment by kiliman** [comment]
Here's a simple diagram explaining the difference.

With cookie-based session storage, **ALL** the session data for that user is stored in the cookie itself. This works well if you have a small amount of session data. 1) The browser typically has a 4KB limit on cookie size. 2) The cookie is sent on **EVERY** request, so a lot of data can mean wasted network data for your user.

With non-cookie-based session storage, the cookie **ONLY** contains a `session id`. This id is used to read/write to the session storage. This is typically file system storage, Redis cache, or a database. Since the cookie only has a session id, this eliminates the issues noted above. The drawback though is you now have to maintain this additional session storage.

<img width="881" alt="image" src="https://user-images.githubusercontent.com/47168/208945242-432c4954-e0a1-4ecd-9b45-22039da57c5f.png">

**Reply by juniorforlife** [reply]
@kiliman Hi, thanks for the answer. But still I don't see how the cookie session is different than a "normal" cookie? If all the session data is stored in the cookie then isn't it the cookie itself?

**Reply by sergiodxa** [reply]
They will essentially be the same since both store the data on the cookie itself, the difference is on the API you use to work with them.

With `createCookie` what you get is a Cookie object with the `parse` and `serialize` method that you need to use to read the cookie value from the Cookie header string or get the value to put in Set-Cookie.

With `createCookieSessionStorage` you get a SessionStorage object with the `getSession`, `commitSession` and `destroySession`, and the `getSession` gives you a Session object.

The nice thing about using sessions here instead of a cookie is that, you could in the future replace `createCookieSessionStorage` with any other way to create a SessionStorage object and the rest of your code will remain unchanged because your code knows how to work with a SessionStorage, doesn't matter if the session data is stored in the cookie or not.

So it's a matter of the API, SessionStorage allows to change the internal mechanism using for storing the session data, if you used a Cookie object instead and later changed to a session storage you would have to refactor many things.

Aside of that, there are other nice things sessions have, like `session.flash` where you could set a one-time read value in your session, but at the end is not impossible to re-create them with Cookie since SessionStorage uses Cookie objects internally.

**Reply by juniorforlife** [reply]
Beautifully explained! Thanks a lot 🙏
